diff --git Makefile Makefile
index 27dacd03..f08ce388 100644
--- Makefile
+++ Makefile
@@ -24,7 +24,7 @@ OBJ_FILES = $(OBJDIR)/*.o
 # profile tool : air
 
 
 #################################################
 # nvme driver : unvme, libaio
diff --git lib/air/Makefile lib/air/Makefile
index 5cd021a8..1ea85144 100644
--- lib/air/Makefile
+++ lib/air/Makefile
@@ -6,7 +6,7 @@ BINDIR = $(TOP)/bin
 OBJDIR = $(BINDIR)/obj
 OBJ_FILES = $(OBJDIR)/*.o
 
