--- src/process/process_manager.h	2019-12-23 14:16:15.929203239 +0900
+++ test/acceptance_test/resource_usage/empty_node/new_process_manager.h	2019-12-23 14:13:54.186303533 +0900
@@ -42,7 +42,7 @@
                     air::ProcessorType ptype);
     void StreamResource(uint32_t node_cnt, uint32_t node_idx,
                         air::ProcessorType ptype);
-    Processor* processor[to_dtype(air::NodeID::COUNT)] {nullptr,};
+    Processor* processor[to_dtype(air::NodeID::COUNT)];
     meta::GlobalMetaGetter* global_meta_getter {nullptr};
     meta::NodeMetaGetter* node_meta_getter {nullptr};
     node::NodeManager* node_manager {nullptr};
